components/places/src/storage/bookmarks.rs

```diff
@@ -121,6 +121,45 @@ pub enum BookmarkPosition {
     Append,
 }
 
+/// Helpers to deal with managing the position correctly.
+
+/// Updates the position of existing items so that the insertion of a child in
+/// the position specified leaves all siblings with the correct position.
+/// Returns the index the item should be inserted at.
+fn resolve_pos_for_insert(
+    db: &impl ConnExt,
+    pos: &BookmarkPosition,
+    parent_id: RowId,
+    cur_child_count: u32,
+) -> Result<u32> {
+    Ok(match pos {
+        BookmarkPosition::Specific(specified) => {
+            let actual = min(*specified, cur_child_count);
+            // must reorder existing children.
+            db.execute_named_cached(
+                "UPDATE moz_bookmarks SET position = position + 1
+                 WHERE parent = :parent_id
+                 AND position >= :position",
+                &[(":parent_id", &parent_id), (":position", &actual)],
+            )?;
+            actual
+        }
+        BookmarkPosition::Append => cur_child_count,
+    })
+}
+
+/// Updates the position of existing items so that the deletion of a child
+/// from the position specified leaves all siblings with the correct position.
+fn update_pos_for_deletion(db: &impl ConnExt, pos: u32, parent_id: RowId) -> Result<()> {
```

Comment:
`resolve_pos_for_insert` and `update_pos_for_deletion` could _almost_ live in the same function, too, since they differ only by the sign. But I'm also OK with keeping them separate if you think it's clearer!

Reply:
I left them alone as that would have made things less clear IMO